### PR TITLE
feature/enum-annotations: Annotations for enum

### PR DESCRIPTION
**Pull request by matepek**
(In progress...)

I was thinking using the `private` as an annotation but I don't like the idea. It's not for that and values will be annotated using 2 lists so the second one should have the `private` or the first but then the second has to be empty. Of course I could switch them but thats not a solution and the problem will be the same.

### CODE REVIEW

File: docs/src/main/markdown/customizations.md
Comment (thesamet) [inline docs/src/main/markdown/customizations.md:724]: 0.11.4

File: protobuf/scalapb/scalapb.proto
Comment (thesamet) [inline protobuf/scalapb/scalapb.proto:300]: typo: Unrecognized :)

File: protobuf/scalapb/scalapb.proto
Comment (thesamet) [inline protobuf/scalapb/scalapb.proto:295]: For consistency with the existing `companion_annotations` name, please rename to`trait_annotations` - same comment apply to the new fields below.

File: protobuf/scalapb/scalapb.proto
Comment (thesamet) [inline protobuf/scalapb/scalapb.proto:294]: s/generated case object/generated enum's base trait/

File: protobuf/scalapb/scalapb.proto
Comment (thesamet) [inline protobuf/scalapb/scalapb.proto:297]: I think it should be "all generated case objects"? Is it including or excluding the Unrecognized - let's specify.

File: protobuf/scalapb/scalapb.proto
Comment (thesamet) [inline protobuf/scalapb/scalapb.proto:303]: I prefer to avoid having specific presets for consistency with the other annotations. There are different types of private in Scala (private and package-private) and different users might want different things.

File: CHANGELOG.md
Comment (thesamet) [inline CHANGELOG.md:5]: Our changelog is out of date. This should go into a new "0.11.4 (Unreleased)" section.

Comment (thesamet) [inline CHANGELOG.md:5]: options for enum annotations

File: compiler-plugin/src/main/scala/scalapb/compiler/DescriptorImplicits.scala
Comment (thesamet) [inline compiler-plugin/src/main/scala/scalapb/compiler/DescriptorImplicits.scala:832]: unrecognizedAnnotations

File: compiler-plugin/src/main/scala/scalapb/compiler/DescriptorImplicits.scala
Comment (thesamet) [inline compiler-plugin/src/main/scala/scalapb/compiler/DescriptorImplicits.scala:835]: Let's remove (see rational in another comment)

File: compiler-plugin/src/main/scala/scalapb/compiler/DescriptorImplicits.scala
Comment (thesamet) [inline compiler-plugin/src/main/scala/scalapb/compiler/DescriptorImplicits.scala:826]: traitAnnotations

File: compiler-plugin/src/main/scala/scalapb/compiler/DescriptorImplicits.scala
Comment (thesamet) [inline compiler-plugin/src/main/scala/scalapb/compiler/DescriptorImplicits.scala:829]: valueAnnotations

File: protobuf/scalapb/scalapb.proto
Comment (thesamet) [inline protobuf/scalapb/scalapb.proto:324]: "to the generated case object for this enum vlaue".

File: compiler-plugin/src/main/scala/scalapb/compiler/DescriptorImplicits.scala
Comment (thesamet) [inline compiler-plugin/src/main/scala/scalapb/compiler/DescriptorImplicits.scala:829]: Shouldn't this method take the EnumValueDescriptor too so it concats the annotations from the EnumValueDescriptor's options?

**Comment by thesamet** [comment]
Thanks for working on this PR. Comments are mostly around naming consistency.  Yes, let's add tests too.

**Review by thesamet — CHANGES_REQUESTED** [review]

File: protobuf/scalapb/scalapb.proto
Comment (matepek) [reply protobuf/scalapb/scalapb.proto:303]: I'm moving this conversation to the main thread

**Comment by matepek** [comment]
@thesamet:
> I prefer to avoid having specific presets for consistency with the other annotations. There are different types of private in Scala (private and package-private) and different users might want different things.

Consider the followings:
- This is a library. Its inner business that something is private or not. With this feature we are just encourage the user to use it correctly.
- Having this private option: the user still can disable private and add some "magic" annotation like `private[package]`
- As far as I understand the intended usage of this lib the users shouldn't create `Unrecognized` values by themselves. My issue shows exactly why. I don't see much point exposing it to them.
- Once again, (because i think this is the most important point) encouraging the user to to use the lib correctly. 

(Maybe I would also consider making it private by default in the next major version.)

What do you think?

File: compiler-plugin/src/main/scala/scalapb/compiler/DescriptorImplicits.scala
Comment (matepek) [reply compiler-plugin/src/main/scala/scalapb/compiler/DescriptorImplicits.scala:829]: I was debating pros and cons and ended up leaving like this because the reader can see it better whats happening in the generator:
```
          .add("@SerialVersionUID(0L)")
          .seq(e.valueAnnotationList)
          .seq(v.annotationList)
```

Another thing I was thinking: Is there a scenario when the order of the annotations really matter? Shouldn't right?

**Comment by thesamet** [comment]
> What do you think?
Overall, I think that making it private could have been a valid design choice when `Unrecognized` were introduced. Making it private now, either in opt-in or opt-out mode, comes at a cost to users if its opt-out and to maintainers if it's opt-in.

For users: the current legit use case by users is pattern matching. Users handle the unrecognized case by matching on `_: Unrecognized`, marking it private by default in the future would break user code. Also, test code out there is instantiating using `Unrecognized` to test scenarios. I agree that it should be preferred to instantiate using `fromValue`.

For maintainers: every option added comes with maintenance cost. Adding an explicit `unrecognized_is_private` option communicates to users that this is a supported and tested combination and it generally gets maintained long term. Given that the feedback you provided on `Unrecognized` constructor being public hasn't been voiced so far, I prefer at this point to not take on supporting it. However, my goal is to provide value to users through this library, so I am willing to reconsider in the future.

**Comment by matepek** [comment]
Pattern matching of course... Good point.

Hm... How about this?

```
   s"""final case class Unrecognized private[$name](unrecognizedValue: _root_.scala.Int) extends $name(unrecognizedValue) with _root_.scalapb.UnrecognizedEnum
           |object Unrecognized {
           |  def apply(unrecognizedValue: _root_.scala.Int): $name = fromValue(unrecognizedValue) 
           |}
           |lazy val values = scala.collection.immutable.Seq(${e.getValues.asScala
          .map(_.scalaName.asSymbol)
          .mkString(", ")})
           |def fromValue(__value: _root_.scala.Int): $name = __value match {""".stripMargin
      )
      .print(e.valuesWithNoDuplicates) { case (p, v) =>
        p.add(s"  case ${v.getNumber} => ${v.scalaName.asSymbol}")
      }
      .add(
        s"""  case __other => new Unrecognized(__other)
           |}
```

**Comment by thesamet** [comment]
I think it's a fairly reasonable change, my earlier comment still applies - though I want to refrain from breaking public API or maintaining an opt-in feature that needs to be tested/supported in the long run.

**Comment by matepek** [comment]
Your concerns are valid. I just believe that this situation should be handled. But the newly proposed one is actually better than the one I previously suggested.

I added `@deprecated` annotation to the Unrecognized.apply.
Because of this change the annotations are not necessary for the purpose but I can continue it.

If there are no further suggestions I will finish it soon.

File: compiler-plugin/src/main/scala/scalapb/compiler/ProtobufGenerator.scala
Comment (thesamet) [inline compiler-plugin/src/main/scala/scalapb/compiler/ProtobufGenerator.scala:76]: That's a great idea, but unfortunately it breaks binary compatibility (see the mima test). I suggest that the changes around this (deprecation and making the constructor private) are split to a different PR which we can target for next stable (0.12 or 1.0)

File: CHANGELOG.md
Comment (thesamet) [inline CHANGELOG.md:5]: This line needs to be updated.

**Comment by matepek** [comment]
Hey there,
So I couldn't figure out in a reasonable time how to write a test which checks the annotation of a scala object so that part is missing (see TODO), otherwise It might ready. pls check.

File: protobuf/scalapb/scalapb.proto
Comment (thesamet) [inline protobuf/scalapb/scalapb.proto:298]: recognized

File: docs/src/main/markdown/customizations.md
Comment (thesamet) [inline docs/src/main/markdown/customizations.md:725]: Can you fix the example? Once merged this gets automatically deployed to the live site.

**Comment by thesamet** [comment]
Reviewed - left some comments. Sorry for the delay, was away and missed the notifications.

File: protobuf/scalapb/scalapb.proto
Comment (matepek) [reply protobuf/scalapb/scalapb.proto:298]: sure

File: docs/src/main/markdown/customizations.md
Comment (matepek) [reply docs/src/main/markdown/customizations.md]: sure

**Pull request merged**
